easybuild/tools/filetools.py

```diff
@@ -1495,19 +1495,45 @@ def create_patch_info(patch_spec):
                                  str(patch_spec))
 
     elif isinstance(patch_spec, string_type):
-        allowed_patch_exts = ['.patch' + x for x in ('',) + ZIPPED_PATCH_EXTS]
-        if not any(patch_spec.endswith(x) for x in allowed_patch_exts):
-            msg = "Use of patch file with filename that doesn't end with correct extension: %s " % patch_spec
-            msg += "(should be any of: %s)" % (', '.join(allowed_patch_exts))
-            _log.deprecated(msg, '5.0')
+        validate_patch_spec(patch_spec)
         patch_info = {'name': patch_spec}
+    elif isinstance(patch_spec, dict):
+        valid_keys = ['name', 'copy', 'level', 'sourcepath', 'alt_location']
+        patch_info = {}
+        for key in patch_spec.keys():
+            if key in valid_keys:
+                patch_info[key] = patch_spec[key]
+            else:
+                raise EasyBuildError("Wrong patch spec '%s', use of unknown key %s in dict (valid keys are %s)",
+                                     str(patch_spec), key, valid_keys)
+
+        # Dict must contain at least the patchfile name
+        if 'name' not in patch_info.keys():
+            raise EasyBuildError("Wrong patch spec '%s', when using a dict 'name' entry must be supplied",
+                                 str(patch_spec))
+        if 'copy' not in patch_info.keys():
+            validate_patch_spec(patch_info['name'])
+        else:
+            if 'sourcepath' in patch_info.keys() or 'level' in patch_info.keys():
+                raise EasyBuildError("Wrong patch spec '%s', you can't use 'sourcepath' or 'level' with 'copy' (since "
+                                     "this implies you want to copy a file to the 'copy' location)",
+                                     str(patch_spec))
     else:
-        error_msg = "Wrong patch spec, should be string of 2-tuple with patch name + argument: %s"
+        error_msg = "Wrong patch spec, should be string, 2-tuple with patch name + argument, or a dict " \
+                    "(with possible keys ['name', 'copy', 'level', 'sourcepath', 'alt_location']): %s"
```

Comment:
This should definitely be derived from `valid_keys` rather than being hardcoded again here...

We're now already hardcoding the valid keys in two places (in `create_patch_info` + `fetch_patches`), which we should also try to avoid (but that's less easy I guess)

Reply:
@ocaisa could you fix the first part of that so we can get this merged.

On the hardcoding part, that would involve defining macros for ALT_LOCATION etc, but I don't see the point really.

Reply:
This should be fixed in [68ade05](https://github.com/easybuilders/easybuild-framework/pull/3994/commits/68ade05bc5a0c7006f74b2e694c288d2287d0c4a)